test/functional/Progress.hs

```diff
@@ -59,47 +76,87 @@ formatLspConfig provider = object ["haskell" .= object ["formattingProvider" .=
       _ -> Nothing)
-    end = EndM <$> satisfyMaybe (\case
+    progressEnd = ProgressEnd <$> satisfyMaybe (\case
       FromServerMess SProgress (NotificationMessage _ _ (ProgressParams t (End x))) -> Just (ProgressParams t x)
       _ -> Nothing)
-    expectElem a as = a `elem` as @? "Unexpected " ++ show a
+
+interestingMessage :: Session a -> Session (InterestingMessage a)
+interestingMessage interestingMessage =
+  fmap InterestingMessage interestingMessage <|> fmap ProgressMessage progressMessage
+
+expectProgressMessagesTill :: Show a => Session a -> [Text] -> [ProgressToken] -> Session (a, [ProgressToken])
+expectProgressMessagesTill stopMessage expectedTitles activeProgressTokens = do
+  message <- skipManyTill anyMessage (interestingMessage stopMessage)
+  case message of
+    InterestingMessage a -> do
+      pure (a, activeProgressTokens)
+    ProgressMessage (ProgressCreate params) -> do
+      expectProgressMessagesTill stopMessage expectedTitles (getToken params : activeProgressTokens)
+    ProgressMessage (ProgressBegin params) -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessagesTill stopMessage (delete (getTitle params) expectedTitles) activeProgressTokens
+    ProgressMessage (ProgressReport params) -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessagesTill stopMessage expectedTitles activeProgressTokens
+    ProgressMessage (ProgressEnd params) -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessagesTill stopMessage expectedTitles (delete (getToken params) activeProgressTokens)
+
+{- | Test that the server is correctly producing a sequence of progress related
+ messages. Each create must be pair with a corresponding begin and end,
+ optionally with some progress in between. Tokens must match. The begin
+ messages have titles describing the work that is in-progress, we check that
+ the titles we see are those we expect.
+-}
+expectProgressMessages :: [Text] -> [ProgressToken] -> Session ()
+expectProgressMessages [] [] = pure ()
+expectProgressMessages expectedTitles activeProgressTokens = do
+  message <- skipManyTill anyMessage progressMessage
+  case message of
+    ProgressCreate params -> do
+      expectProgressMessages expectedTitles (getToken params : activeProgressTokens)
+    ProgressBegin params -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessages (delete (getTitle params) expectedTitles) activeProgressTokens
+    ProgressReport params -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessages expectedTitles activeProgressTokens
+    ProgressEnd params -> do
+      liftIO $ getToken params `expectedIn` activeProgressTokens
+      expectProgressMessages expectedTitles (delete (getToken params) activeProgressTokens)
```

Comment:
Can `expectProgressMessages` be defined in terms of `expectProgressMessagesTill`?

Reply:
I couldn't figure out how to write one in terms of the other because the termination conditions differ slightly, but did the next best thing of factoring out the common logic.